File: COURSE 1 (Algorithmic Toolbox)/week5_dynamic_programming1/Week 5 Assignment/submitted files C++/change_dp.cpp
#include <iostream>
#include <vector>
using namespace std;

int dpchange(int total) {
    vector<int> options{1,3,4};
    int minnumofop[total + 1];
    minnumofop[0] = 0;
    int num;
    for (int i = 1; i <= total; i++) {
        minnumofop[i] = 2147483647;
        for (int j = 0; j < options.size(); j++) {
            if (i >= options[j]) {
                num = minnumofop[i - options[j]] + 1;
                if (num < minnumofop[i]) {
                    minnumofop[i] = num;
                }
            }
        }
    }
    return minnumofop[total];
}

int main() {
  int m;
  std::cin >> m;
  std::cout << dpchange(m) << '\n';
}

File: COURSE 1 (Algorithmic Toolbox)/week5_dynamic_programming1/Week 5 Assignment/src/com/company/ChangeDP.java
package com.company;

import java.util.Scanner;

public class ChangeDP {

    public static int dpChange(int[] options, int total) {
        int[] minNumOfOp = new int[total + 1];
        minNumOfOp[0] = 0;
        int num;
        for (int i = 1; i <= total; i++) {
            minNumOfOp[i] = Integer.MAX_VALUE;
            for (int j = 0; j < options.length; j++) {
                if (i >= options[j]) {
                    num = minNumOfOp[i - options[j]] + 1;
                    if (num < minNumOfOp[i]) {
                        minNumOfOp[i] = num;
                    }
                }
            }
        }
        return minNumOfOp[total];
    }

    public static void main(String[] args) {
        Scanner scanner=new Scanner(System.in);
        System.out.println(dpChange(new int[]{1, 3, 4}, scanner.nextInt()));
    }
}


